muzzle/src/main/java/io/opentelemetry/javaagent/tooling/HelperInjector.java

```diff
@@ -124,38 +125,41 @@ public HelperInjector(
     this.instrumentation = instrumentation;
   }
 
-  private HelperInjector(
+  public HelperInjector(
       String requestingName,
-      Map<String, Supplier<byte[]>> helperMap,
+      Map<String, Function<ClassLoader, byte[]>> helperMap,
+      List<HelperResource> helperResources,
+      ClassLoader helpersSource,
       Instrumentation instrumentation) {
     this.requestingName = requestingName;
 
     this.helperClassNames = helperMap.keySet();
     this.dynamicTypeMap.putAll(helperMap);
 
-    this.helperResources = Collections.emptyList();
-    this.helpersSource = null;
+    this.helperResources = helperResources;
+    this.helpersSource = helpersSource;
     this.instrumentation = instrumentation;
   }
 
   public static HelperInjector forDynamicTypes(
       String requestingName,
       Collection<DynamicType.Unloaded<?>> helpers,
       Instrumentation instrumentation) {
-    Map<String, Supplier<byte[]>> bytes = new HashMap<>(helpers.size());
+    Map<String, Function<ClassLoader, byte[]>> bytes = new HashMap<>(helpers.size());
     for (DynamicType.Unloaded<?> helper : helpers) {
-      bytes.put(helper.getTypeDescription().getName(), helper::getBytes);
+      bytes.put(helper.getTypeDescription().getName(), cl -> helper.getBytes());
     }
-    return new HelperInjector(requestingName, bytes, instrumentation);
+    return new HelperInjector(
+        requestingName, bytes, Collections.emptyList(), null, instrumentation);
   }
 
   public static void setHelperInjectorListener(HelperInjectorListener listener) {
     helperInjectorListener = listener;
   }
 
-  private Map<String, Supplier<byte[]>> getHelperMap() {
+  private Map<String, Supplier<byte[]>> getHelperMap(ClassLoader targetClassloader) {
```

Comment:
Passing the class loader here feels unnecessary. As far as I understand the class loader you need here is either agent class loader or extension class loader, it is ok to keep reference to these, they won't go away anyway. If you are currently using instrumentation module class loader here then consider whether your really need to. If you need instrumentation class loader because the injected class may reference some type that is only present in the application and byte-buddy blows up when it is not available then perhaps it would make more sense to generate the proxy with asm.

Reply:
>  If you need instrumentation class loader because the injected class may reference some type that is only present in the application and byte-buddy blows up when it is not available then perhaps it would make more sense to generate the proxy with asm.

That is precisely the case, I've even ensured that this is covered by the tests: The `MyProxy` class extends `MyProxySuperclass` which is only available in the instrumented application classloader.

We could remove the need for having the classloader at hand when generating the proxies, but I went for the other route for the following reasons:
 * I would assume that the `ByteBuddy`-based proxy generator is a lot easier to read and maintain than an ASM one. I'd expect the ASM proxy generator to be more brittle.
 * This would not allow us to override & delegate inherited methods by the proxy. We don't do this currently, but I would suspect it could become relevant in the future. See also [this test-case](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d0523fd9e9ed185ceddc096a9e2e16939083871a/javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyProxyFactoryTest.java#L276) for reference.
 
 What problems do you see with providing the instrumented app classloader when injecting the classes?